plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -606,25 +609,81 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
             throws IOException
     {
-        try {
-            if (!directory(path)) {
-                return deleteObject(keyFromPath(path));
+        String key = keyFromPath(path);
+        if (recursive) {
+            DeletePrefixResult deletePrefixResult;
+            try {
+                deletePrefixResult = deletePrefix(path);
             }
+            catch (AmazonClientException e) {
+                throw new IOException("Failed to delete paths with the prefix path " + path, e);
+            }
+            if (deletePrefixResult == DeletePrefixResult.NO_KEYS_FOUND) {
+                // If the provided key is not a "directory" prefix, attempt to delete the object with the specified key
+                deleteObject(key);
+            }
+            else if (deletePrefixResult == DeletePrefixResult.DELETE_KEYS_FAILURE) {
+                return false;
+            }
+            deleteObject(key + DIRECTORY_SUFFIX);
         }
-        catch (FileNotFoundException e) {
-            return false;
+        else {
+            Iterator<ListObjectsV2Result> listingsIterator = listObjects(path, OptionalInt.of(2), true);
+            Iterator<String> objectKeysIterator = Iterators.concat(Iterators.transform(listingsIterator, TrinoS3FileSystem::keysFromRecursiveListing));
+            if (objectKeysIterator.hasNext()) {
+                String childKey = objectKeysIterator.next();
+                if (!Objects.equals(childKey, key + PATH_SEPARATOR) || objectKeysIterator.hasNext()) {
+                    throw new IOException("Directory " + path + " is not empty");
+                }
+                deleteObject(childKey);
+            }
+            else {
+                // Avoid deleting the bucket in case that the provided path points to the bucket root
```

Comment:
I don't think we would delete a bucket, would we?

Maybe 
_Do not attempt to "delete the bucket", of the path is the bucket root, as this would fail_

Reply:
is it true for S3, or another MinIO quirk?

Reply:
I just tried an adaptation of the test `io.trino.plugin.hive.s3.TestTrinoS3FileSystemMinio.testDeleteNonRecursivelyEmptyBucketRoot` on an AWS S3  while having the check for the empty key commented out and it turns out that also on AWS S3 the bucket is being removed.

> is it true for S3, or another MinIO quirk?

Yes, it is true for both `S3` and `MinIO`.